feat: Add fine-grained permission control and category resolution times OP-2804

anthbel opened the pull request:
## Summary
Enhances the grievance module with fine-grained access control at category/flag level and adds support for category-specific resolution times.

## Key Features
- **Permission-based access control**: Restrict categories/flags using permission
- **Category resolution times**: Define resolution times per category with inheritance
- **Hierarchical categories**: Support parent-child category relationships
- **Automatic filtering**: Ticket queries automatically filtered by user permissions
- **Unified configuration**: Support both simple strings and enhanced objects in config

## Configuration Example
```json
{
  "grievance_types": [{
    "name": "complaint",
    "permissions": ["127001", "127002"],
    "resolution_times": "3,0",
    "children": [{
      "name": "urgent",
      "resolution_times": "1,0"
    }]
  }]
}
```
## Backward Compatibility
Fully backward compatible - existing configurations continue to work unchanged

anthbel committed:
feat: Add fine-grained permission control and category resolution times
- Implement permission-based access control for categories/flags
- Add category-specific resolution times with inheritance support
- Create hierarchical category structure with parent-child relationships
- Add automatic ticket filtering based on user permissions
- Support unified configuration format for both simple and enhanced setups

anthbel commented:
Hi @weilu, @Shahzaibahmad97 - This Draft PR is ready for your review. Looking forward to your feedback.

weilu commented on README.md:
````diff
@@ -2,12 +2,165 @@
+```json
+{
+  "resolution_times": "5,0",  // Global default: 5 days
+  "default_resolution": {     // Legacy configuration (still supported)
+    "feedback": "7,0"
+  },
````
I'd suggest we remove this legacy configuration for 25 or 26 major release so we don't have to maintain two sets of parallel config options going forward.

delcroip replied:
I think it might be better to foresee a migration of the config rather than supporting the legacy version

anthbel replied:
Migration will have to be manual. We can remove it and still handle before the next major release.
```suggestion
```

weilu commented on README.md:
```diff
@@ -2,12 +2,165 @@
+
+### Flags Configuration
+
+Similarly, flags support both simple and enhanced formats:
```
Does flags config also support nested parent-child options like grievance types? It's unclear from the readme description and examples here.

anthbel replied:
No, I will make it explicit in the doc.

weilu commented on README.md:
```diff
@@ -2,12 +2,165 @@
+    {
+      "name": "sensitive",
+      "priority": "Critical",
+      "permissions":["127001", "127002"]
```
I don't know how useful this permissions configuration is in generalized cases. For Malawi public works case, there are two case permission control requirements:

1. district vs. national case manager: national case manager user role should be able to see all cases, while district case managers can only see cases from their assigned district. There's existing patterns for handling this in openimis, which is known as (location-based) row level security. Here are examples for [query](https://github.com/openimis/openimis-be-individual_py/blob/develop/individual/models.py#L36-L62) & [mutation](https://github.com/openimis/openimis-be-individual_py/blob/develop/individual/gql_mutations.py#L94-L101), and the config is `settings.ROW_SECURITY`. User's location permission can be configured at [district level](https://github.com/openimis/openimis-be-location_py/blob/develop/location/models.py#L571) or [village level](https://github.com/openimis/openimis-be-location_py/blob/develop/location/models.py#L689), but in practice I think only district level is used for row-level security (please double check).
2. Only National Gender Officer can view and manage GBV cases: this would require permission control at the grievance type level. Let me know your thoughts on how this can be best implemented @anthbel. Happy to brainstorm too.

delcroip replied:
the permission are multi factored:
- location based
- authority (the permission you have here)
- business defined : look for a link between the user and the resource


for more flexibility, business defined  might be more suitable here, ex:
- creating a resource called "Grievance sensitiveness" on which we could attach role
- upon grievance access we could check that if there is a sensitiveness defined, the user belong to an authorized role

delcroip replied:
again I would rather have roles instead of permission, it create more flexibility; in other case we managed that with real resource but config base is cheaper and should work fine

weilu reviewed (commented):
@anthbel I'm still reviewing the rest of this PR but just wanted to flag on the permission control part. Let me know what you think, both in terms of use cases and implementation.

anthbel committed:
feat: Implement auto-generated permissions and field-level access control
Permission system auto-generates IDs for categories/flags

  Major enhancements:
  - Add automatic permission ID generation (127100-127999 range) for categories/flags
  - Implement three-tier access control: restricted_read, read, and full access
  - Add field-level security with configurable visible_fields per category
  - Enforce security at Django model level via get_queryset override

  Configuration enhancements:
  - Support simplified permission format: ["read", "create", "update"]
  - Categories inherit parent permissions and visible_fields
  - Children can only make permissions MORE restrictive than parents
  - Backward compatible with existing numeric permission IDs

  New features:
  - Management command supports list, export, cleanup operations
  - Permissions stored persistently in Django auth_permission table
  - Stable permission IDs maintained across configuration changes

weilu commented on grievance_social_protection/tests/test_config_processing.py:
```diff
@@ -0,0 +1,213 @@
+                {
+                    'name': 'complex',
+                    'priority': 'High',
+                    'permissions': ['127001', '127002'] 
```
minor: consider making this and below test config examples consistent with actual use cases and use 'read', 'create' etc instead of numbers.

weilu commented on grievance_social_protection/apps.py:
```diff
@@ -91,34 +126,236 @@ def get_grievance_type_options_msg(types):
+                # Simple string format (backward compatible)
+                full_name = f"{parent_name}|{item}" if parent_name else item
+                processed_categories[full_name] = {
+                    'priority': parent_info.get('priority', 'Medium') if parent_info else 'Medium',
```
Consider simplification: `(parent_info or {}).get('priority', 'Medium')` Ditto for below

weilu commented on grievance_social_protection/apps.py:
```diff
@@ -91,34 +126,236 @@ def get_grievance_type_options_msg(types):
+                        logger.info(f"Auto-adding 'restricted_read' permission to category '{cat_name}' due to visible_fields")
+                    if 'read' not in permissions:
+                        permissions.append('read')
+                        logger.info(f"Auto-adding 'read' permission to category '{cat_name}' due to visible_fields")
```
Consider refactoring to keep it DRY:

```
for perm in ['restricted_read', 'read']:
  if perm not in permissions:
    permissions.append(perm)
    logger.info(f"Auto-adding '{perm}' permission to category '{cat_name}' due to visible_fields")
```

weilu commented on grievance_social_protection/apps.py:
```diff
@@ -91,34 +126,236 @@ def get_grievance_type_options_msg(types):
+                permissions = flag.get('permissions', [])
+                # Convert dict format to list format if needed
+                if isinstance(permissions, dict):
+                    permissions = list(permissions.keys())
```
Why would permissions be in dict format?

weilu commented on grievance_social_protection/rights.py:
```diff
@@ -0,0 +1,207 @@
+            ct = ContentType.objects.get_for_model(Ticket)
+        except Exception as e:
+            logger.warning(f"Could not get ContentType for Ticket model: {e}")
+            return
```
I'd suggest leaving this Exception unhandled as if it were to happen it's something fundamentally wrong and we want to "fail fast"

weilu commented on grievance_social_protection/rights.py:
```diff
@@ -0,0 +1,207 @@
+        # Get used IDs in our range
+        used_ids = set(Permission.objects.filter(
+            id__range=(cls.GRIEVANCE_RIGHT_BASE, cls.GRIEVANCE_RIGHT_MAX)
+        ).values_list('id', flat=True))
```
Can we reuse existing_perms instead of creating a new variable here?

anthbel replied:
We cannot reuse existing_perms directly because the IDs in existing_perms represent only a subset of used_ids.
The existing_perms contains permissions created by the automatic permission.
However, used_ids includes all occupied IDs within the fixed range, that can include system-generated permissions that we don't control.

weilu commented on grievance_social_protection/rights.py:
```diff
@@ -0,0 +1,207 @@
+            if permissions:
+                # Convert back to list format if still dict
+                if isinstance(permissions, dict):
+                    permissions = list(permissions.keys())
```
Again, I'm not sure under what circumstances would permissions be a dict

weilu commented on grievance_social_protection/rights.py:
```diff
@@ -0,0 +1,207 @@
+        
+        for category_name, category_info in processed_categories.items():
+            permissions = category_info.get('permissions', [])
+            if permissions:
```
because of the for loop starting on line 63, this if check isn't necessary. Consider removing.

weilu commented on grievance_social_protection/rights.py:
```diff
@@ -0,0 +1,207 @@
+            'restricted_read': 4,   # -> gql_query_restricted_*
+        }
+        
+        suffix = suffix_map.get(perm_type, 9)
```
Consider `suffix_map[perm_type]` which would force invalid perm_type to cause a KeyError instead of generating an invalid id that ends with 9.

weilu commented on grievance_social_protection/rights.py:
```diff
@@ -0,0 +1,207 @@
+                            logger.info(f"Created new permission: {codename} (ID: {new_id})")
+                        except ValueError as e:
+                            logger.error(f"Failed to create permission {codename}: {e}")
+                            continue
```
When a permission failed to be created due to IDs are exhausted should we consider not catch the error and just let it throw there? Additionally, should we consider wrapping all Permission generations in the for loop in a transaction so either they all succeed or fail together? This way if one or a few were to fail we just clean up and redo the whole batch?

weilu commented on grievance_social_protection/rights.py:
```diff
@@ -0,0 +1,207 @@
+                    elif perm_type == 'delete':
+                        right_name = f"gql_mutation_delete_{safe_name}_category_tickets_perms"
+                    else:
+                        continue
```
Consider refactoring:

```
prefix_map = {
    'restricted_read': 'gql_query_restricted',
    'read': 'gql_query',
    'create': 'gql_mutation_create',
    'update': 'gql_mutation_update',
    'delete': 'gql_mutation_delete',
}

prefix = prefix_map[perm_type]
right_name = f"{prefix}_{safe_name}_category_tickets_perms"
```

Additionally, this way invalid perm_type will raise an exception and lead user to fix config

weilu commented on grievance_social_protection/rights.py:
```diff
@@ -0,0 +1,207 @@
+                    elif perm_type == 'delete':
+                        right_name = f"gql_mutation_delete_{safe_name}_flagged_tickets_perms"
+                    else:
+                        continue
```
This logic is very similar to that above. Consider refactor to keep it DRY. Also all comments above apply here.

weilu commented on grievance_social_protection/models.py:
```diff
@@ -61,7 +62,12 @@ def __str__(self):
+        queryset = queryset.filter(
+            models.Q(date_valid_from__lte=TimeUtils.now()) | models.Q(date_valid_from__isnull=True),
+            models.Q(date_valid_to__gte=TimeUtils.now()) | models.Q(date_valid_to__isnull=True),
+        )
```
@anthbel what's the issue with using `core.filter_validity` here?

weilu commented on grievance_social_protection/models.py:
```diff
@@ -109,6 +116,32 @@ def clean(self):
                 raise ValueError("Another comment for this ticket is already marked as resolved.")
 
+    @classmethod
+    def filter_queryset(cls, queryset=None):
+        if queryset is None:
+            queryset = cls.objects.all()
+        queryset = queryset.filter(is_deleted=False)
+        return queryset
```
Would this need filtering by validity?

weilu commented on grievance_social_protection/tests/test_ticket_service_permissions.py:
```diff
@@ -0,0 +1,310 @@
+        from grievance_social_protection.rights import GrievanceRightsManager
+        
+        # Generate rights using the real rights manager
+        GrievanceRightsManager.generate_automatic_rights(TicketConfig)
```
`GrievanceRightsManager.generate_automatic_rights` is used in test for setup but not actually covered by tests. It would be good to have coverage as the implementation has complexity.

weilu commented on grievance_social_protection/apps.py:
```diff
@@ -91,34 +126,236 @@ def get_grievance_type_options_msg(types):
+        for category_name, category_info in processed_categories.items():
+            resolution_time = category_info.get('resolution_times')
+            if resolution_time:
+                cls.__validate_resolution_time_format(resolution_time, f"category '{category_name}'")
```
The returned boolean value from `__validate_resolution_time_format` isn't used her nor above. I see the validation is only called when resolution_time is explicitly configured. In the same vein as some of the previous comments on exception handling, I'm leaning towards throwing an error and fail dramatically if there's an invalid configuration instead of just logging the issue here, because the goal is to flag to the user when there's an invalid configuration and have the user correct it as soon as possible.

weilu commented on grievance_social_protection/apps.py:
```diff
@@ -91,34 +126,236 @@ def get_grievance_type_options_msg(types):
+    def __validate_resolution_time_format(cls, value, context=""):
+        """Validate a single resolution time format"""
+        if ',' not in value:
+            logger.warning(f"Invalid resolution time format for {context}. "
```
See comment above, consider raising exception instead of logging a warning. Same below.

weilu commented on grievance_social_protection/gql_queries.py:
```diff
@@ -36,20 +36,69 @@ class TicketGQLType(DjangoObjectType):
+                default_hidden = ['description', 'resolution', 'reporter_id', 'channel', 
+                                'attending_staff', 'json_ext', 'category', 'flags', 'title']
+                return field_name in default_hidden
+            # If visible_fields is defined, only show fields in the list
```
This branch is unreachable because `get_visible_fields` already checks for user's access level and returns the default visible fields when it's restricted. Consider remove

weilu replied:
In fact, `_should_restrict_field` can be largely simplified because `get_visible_fields` already checks for user's access level. It's basically:

```
def _should_restrict_field(field_name, root, info):
  visible_fields = GrievanceAccessControl.get_visible_fields(user, root.category)
  if visible_fields is None:
    return False

  return field_name not in visible_fields
```

anthbel replied:
get_visible_fields is based on fields visibility configured for the category, we also need to check for flag related restrictions

weilu commented on grievance_social_protection/gql_queries.py:
```diff
@@ -105,9 +163,97 @@ def resolve_reporter_dob(root, info):
+    def resolve_description(self, info):
+        """Return [Restricted] for restricted access"""
+        if TicketGQLType._should_restrict_field('description', self, info):
+            return "[Restricted]" if self.description else None
```
I'd suggest `return "[Restricted]"` regardless of self.description is None or not.

weilu commented on grievance_social_protection/gql_queries.py:
```diff
@@ -105,9 +163,97 @@ def resolve_reporter_dob(root, info):
+    def resolve_resolution(self, info):
+        """Return [Restricted] for restricted access"""
+        if TicketGQLType._should_restrict_field('resolution', self, info):
+            return "[Restricted]" if self.resolution else None
```
Ditto

weilu commented on grievance_social_protection/gql_queries.py:
```diff
@@ -60,6 +109,9 @@ def resolve_is_history(root, info):
         check_ticket_perms(info)
+        access_level = GrievanceAccessControl.get_user_access_level(info.context.user, root.category, root.flags)
+        if access_level in ['none', 'restricted']:
+            return None  # Hide personal info for restricted access
```
Should this return '[Restricted]' to be consistent with the permission configurable fields below?

weilu commented on grievance_social_protection/gql_queries.py:
```diff
@@ -76,6 +128,9 @@ def resolve_reporter_first_name(root, info):
         check_ticket_perms(info)
+        access_level = GrievanceAccessControl.get_user_access_level(info.context.user, root.category, root.flags)
+        if access_level in ['none', 'restricted']:
+            return None  # Hide personal info for restricted access
```
Ditto

weilu commented on grievance_social_protection/gql_queries.py:
```diff
@@ -105,9 +163,97 @@ def resolve_reporter_dob(root, info):
+    def resolve_channel(self, info):
+        """Return [Restricted] for restricted access"""
+        if TicketGQLType._should_restrict_field('channel', self, info):
+            return "[Restricted]" if self.channel else None
```
return "[Restricted]" regardless

weilu commented on grievance_social_protection/gql_queries.py:
```diff
@@ -105,9 +163,97 @@ def resolve_reporter_dob(root, info):
+    def resolve_category(self, info):
+        """Return category value based on visible_fields"""
+        if TicketGQLType._should_restrict_field('category', self, info):
+            return "[Restricted]" if self.category else None
```
return "[Restricted]" regardless, similarly below

weilu commented on grievance_social_protection/gql_queries.py:
```diff
@@ -105,9 +163,97 @@ def resolve_reporter_dob(root, info):
+    def resolve_flags(self, info):
+        """Return flags value based on visible_fields"""
+        if TicketGQLType._should_restrict_field('flags', self, info):
+            return "[Restricted]" if self.flags else None
```
Consider extracting "[Restricted]" into a constant due to repeated use

weilu commented on grievance_social_protection/gql_queries.py:
```diff
@@ -251,21 +397,92 @@ class ResolutionTimesByCategoryGQLType(ObjectType):
+    grievance_categories_hierarchical = graphene.List(GrievanceCategoryGQLType)
+    grievance_flags_detailed = graphene.List(GrievanceFlagGQLType)
+    accessible_categories = graphene.List(graphene.String)
+    accessible_flags = graphene.List(graphene.String)
```
@anthbel I'm not very familiar with the grievance module – can you kindly remind me where these fields will be used/displayed on UI?

weilu commented on grievance_social_protection/access_control.py:
```diff
@@ -0,0 +1,453 @@
+            update_access = cls.check_category_access(user, category_name, 'update')
+            delete_access = cls.check_category_access(user, category_name, 'delete')
+            read_access = cls.check_category_access(user, category_name, 'read')
+            restricted_access = cls.check_category_access(user, category_name, 'restricted_read')
```
I'd suggest making all these string values 'create', 'update', 'delete' etc constant variables

weilu commented on grievance_social_protection/access_control.py:
```diff
@@ -0,0 +1,453 @@
+                category_access = None  # No restrictions
+            elif create_access or update_access or delete_access:
+                # If user has any write permission, they have full access
+                category_access = 'full'
```
What's the difference between `category_access = None` and `category_access = 'full'`?

weilu commented on grievance_social_protection/access_control.py:
```diff
@@ -0,0 +1,453 @@
+                # For each flag, determine the actual access level
+                if create_access or update_access or delete_access:
+                    # If user has any write permission, they have full access
+                    flag_accesses.append('full')
```
Similarly, what's the difference between "full" and None value for flag access?

weilu commented on grievance_social_protection/access_control.py:
```diff
@@ -0,0 +1,453 @@
+            return False
+        
+        # Use the built-in has_perm method from User model
+        return user.has_perm(str(required_right))
```
The two methods above can be refactored to keep it DRY:

```
@classmethod
def _check_access(cls, user, name, access_type, config_attr):
    """
    Generic access checker for category or flag.

    Args:
        user: Django user object
        name: Category or flag name
        access_type: Type of access ('restricted_read', 'read', 'write', 'update')
        config_attr: 'processed_categories' or 'processed_flags'

    Returns:
        bool: True if user has the requested access, False if denied
    """
    if not user or user.is_anonymous:
        return False

    processed = getattr(TicketConfig, config_attr, {})
    if not processed or name not in processed:
        return True

    info = processed[name]
    required_right = info.get('generated_rights', {}).get(access_type)

    if not required_right:
        return False

    return user.has_perm(str(required_right))


@classmethod
def check_category_access(cls, user, category_name, access_type='read'):
    return cls._check_access(user, category_name, access_type, 'processed_categories')


@classmethod
def check_flag_access(cls, user, flag_name, access_type='read'):
    return cls._check_access(user, flag_name, access_type, 'processed_flags')
```

weilu commented on grievance_social_protection/access_control.py:
```diff
@@ -0,0 +1,453 @@
+            return flag_access
+            
+        # No restrictions on anything
+        return 'full'
```
Consider simplifying the method implementation:

```
@classmethod
def get_user_access_level(cls, user, category_name=None, flag_names=None):
    """
    Determine the highest level of access a user has.

    Returns:
        str: 'full', 'read', 'restricted', or 'none'
    """

    def evaluate_access(check_func, name):
        levels = {
            'full': any(check_func(user, name, t) for t in ('create', 'update', 'delete')),
            'read': check_func(user, name, 'read'),
            'restricted': check_func(user, name, 'restricted_read'),
        }

        if levels['full']:
            return 'full'
        elif levels['read']:
            return 'read'
        elif levels['restricted']:
            return 'restricted'
        elif all(v is False for v in levels.values()):
            return 'none'
        else:
            return None  # No restrictions

    def min_access_level(levels):
        # Return most restrictive among present levels
        for level in ('none', 'restricted', 'read', 'full'):
            if level in levels:
                return level
        return None

    category_access = evaluate_access(cls.check_category_access, category_name) if category_name else None

    flag_accesses = []
    if flag_names:
        for flag in (flag_names.split() if isinstance(flag_names, str) else flag_names):
            level = evaluate_access(cls.check_flag_access, flag)
            if level is not None:
                flag_accesses.append(level)

    flag_access = min_access_level(flag_accesses) if flag_accesses else None

    # Most restrictive applies
    return min_access_level([category_access, flag_access])
```

weilu commented on grievance_social_protection/access_control.py:
```diff
@@ -0,0 +1,453 @@
+        processed_flags = getattr(TicketConfig, 'processed_flags', {})
+        if not processed_flags or flag_name not in processed_flags:
+            # No restrictions defined - return None to indicate no restrictions
+            return None
```
Why do we need to differentiate None and True? If there's no good reason to so I'd suggest to stick with True/False and get rid of None because it's easy for people to skip explicit checking of None if they didn't read into this method before use.

weilu commented on grievance_social_protection/access_control.py:
```diff
@@ -0,0 +1,453 @@
+        return 'full'
+    
+    @classmethod
+    def get_accessible_categories(cls, user, min_access='restricted_read'):
```
I'd suggest renaming this to `get_min_accessible_categories` and get rid of the min_access argument and just hard code 'restricted_read' because that's the only use case so far.

weilu commented on grievance_social_protection/access_control.py:
```diff
@@ -0,0 +1,453 @@
+        for category_name in all_categories:
+            access = cls.check_category_access(user, category_name, min_access)
+            if access is True or access is None:  # True = has access, None = no restrictions
+                accessible_categories.append(category_name)
```
This check is awkward. If `check_category_access` were to only return boolean this would have been 

```
has_access = cls.check_category_access(user, category_name, min_access)
accessible_categories.append(category_name) if has_access
```

which reads better too

weilu commented on grievance_social_protection/access_control.py:
```diff
@@ -0,0 +1,453 @@
+        
+        for flag_name in all_flags:
+            access = cls.check_flag_access(user, flag_name, min_access)
+            if access is True or access is None:  # True = has access, None = no restrictions
```
ditto

weilu commented on grievance_social_protection/access_control.py:
```diff
@@ -0,0 +1,453 @@
+        # Check category access
+        if category:
+            access = cls.check_category_access(user, category, access_type)
+            if access is False:  # Explicitly denied
```
ditto, this would have been `if not access`

weilu commented on grievance_social_protection/access_control.py:
```diff
@@ -0,0 +1,453 @@
+        accessible_categories = cls.get_accessible_categories(user, 'restricted_read')
+        
+        # Filter tickets by accessible categories
+        if accessible_categories is not None:
```
`accessible_categories` would never be None given how it's implemented. Consider remove

weilu commented on grievance_social_protection/schema.py:
```diff
@@ -54,9 +54,14 @@ def resolve_comments(self, info, **kwargs):
+        query = Ticket.objects.filter(*append_validity_filter(**kwargs)).all().order_by('ticket_title', )
+        
+        # Apply category and flag permission filtering
+        from .access_control import GrievanceAccessControl
```
Why import here instead of on top?

weilu commented on grievance_social_protection/services.py:
```diff
@@ -84,6 +89,43 @@ def _generate_code(self, obj_data):
+        flags = obj_data.get('flags')
+        
+        try:
+            GrievanceAccessControl.validate_ticket_access(self.user, category, flags)
```
Did you forget to pass `access_type` argument here? Would also need to take `access_type` as an argument in method signature for `_validate_access_control`

weilu commented on grievance_social_protection/services.py:
```diff
@@ -84,6 +89,43 @@ def _generate_code(self, obj_data):
+        try:
+            GrievanceAccessControl.validate_ticket_access(self.user, category, flags)
+        except PermissionDenied as e:
+            raise ValidationError(str(e))
```
Are you sure `PermissionDenied` needs to be caught here? I've seen it being explicitly raised in schema.py before e.g. https://github.com/openimis/openimis-be-social_protection_py/blob/develop/social_protection/schema.py#L186

anthbel committed:
refactor: update access control to return only boolean values and fix permission generation
  - Remove None return values from access control methods - return True/False
  - Update get_user_access_level() to properly handle categories/flags without restrictions
  - Clean up GrievanceRightsManager with better constants and helper methods organization

anthbel committed:
feat: add default 'uncategorized' grievance type and update tests
  - Add default_grievance_type to TicketConfig with 'uncategorized' as default value
  - Set 'uncategorized' as default category for tickets without explicit category
  - Update config processing to insert default type with ['read', 'update'] permissions

anthbel committed:
Fix: Add generated roles to app config

weilu commented on grievance_social_protection/models.py:
```diff
@@ -56,6 +57,12 @@ def clean(self):
+    def save(self, *args, **kwargs):
+        # Set default category if empty
+        if not self.category:
+            self.category = TicketConfig.default_grievance_type
```
Do we need to handle when category is a dict vs string here?

anthbel replied:
No, TicketConfig.default_grievance_type is set to category name. I'll update the related README section for clarification.

Shahzaibahmad97 committed:
Update rights generation condition in apps.py

Updated rights generation to ignore if NO_DATABASE=True flag is set at deployment steps

Shahzaibahmad97 committed:
Add import for os module in apps.py

anthbel committed:
refactor: code simplification and test improvements
  - Extract helpers and constants for better readability
  - Improve exception handling
  - Add GrievanceRightsManager test coverage
  - Update documentation

anthbel committed:
refactor:
- make TicketGQLType resolvers consistently static methods
- Extract helpers to reduce duplication across services, queries, and management command
- Remove dict-format permission checks

anthbel committed:
refactor:
- Add test helpers
- Restricted filters
- Access control perm fallback

anthbel committed:
Merge branch 'develop' into Allow-grievance-case-to-be-limited-access-to-specific-user-roles-per-config

# Conflicts:
#	grievance_social_protection/tests/test_helpers.py

anthbel committed:
refactor: move inline imports to file top

anthbel committed:
refactor: Raising an exception instead of logging a warning

anthbel committed:
fix: align default visible_fields to include category and priority

anthbel committed:
refactor: add access type constants and fix permission label for hierarchical categories

zikani03 commented:
@weilu please help reviewing this, I can see the dependent PR was approved: https://github.com/openimis/openimis-fe-grievance_social_protection_js/pull/25 - but need this to be looked at before that can be merged. Thanks

weilu commented:
@anthbel please fix module CI, flake8 and address copilot comments

anthbel committed:
Clean flake8 errors

anthbel committed:
refactor: fix Flake and CI

anthbel committed:
refactor:  Code refactoring

anthbel committed:
refactor: code refactoring

anthbel committed:
refactor: code refactoring

anthbel committed:
refactor: code refactoring

anthbel committed:
refactor: code refactoring

anthbel commented on grievance_social_protection/apps.py:
```diff
@@ -91,34 +137,270 @@ def get_grievance_type_options_msg(types):
+
+            type_exists = any(get_category_name(cat) == default_grievance_type for cat in categories)
+
+            # If not found, add it with default permissions so uncategorized
+            # tickets are subject to access control like all other categories.
+            if not type_exists:
+                categories.insert(0, {
+                    'name': default_grievance_type,
+                    'permissions': ['read', 'update']
+                })
+                logger.info(
+                    f"Added default grievance type '{default_grievance_type}' with permissions ['read', 'update']")
```
Intended behavior

anthbel commented on grievance_social_protection/apps.py:
```diff
@@ -91,34 +137,270 @@ def get_grievance_type_options_msg(types):
+            for key in dict_field:
+                value = dict_field[key]
+                if value in ['', None]:
+                    raise ValueError(
+                        f"'{key}' in 'default_resolution' has no value. "
+                        f"Expected format: 'days,hours' (e.g. '{DEFAULT_TIME_RESOLUTION}')."
+                    )
```
Intended behavior

anthbel committed:
refactor: code refactoring

anthbel committed:
refactor: code refactoring

anthbel committed:
refactor: code refactoring

anthbel committed:
refactor: code refactoring

anthbel commented on grievance_social_protection/apps.py:
```diff
@@ -91,34 +138,270 @@ def get_grievance_type_options_msg(types):
+
+            type_exists = any(get_category_name(cat) == default_grievance_type for cat in categories)
+
+            # If not found, add it with default permissions so uncategorized
+            # tickets are subject to access control like all other categories.
+            if not type_exists:
+                categories.insert(0, {
+                    'name': default_grievance_type,
+                    'permissions': ['read', 'update']
+                })
+                logger.info(
+                    f"Added default grievance type '{default_grievance_type}' with permissions ['read', 'update']")
```
Intended behavior

anthbel commented on grievance_social_protection/apps.py:
```diff
@@ -91,34 +138,270 @@ def get_grievance_type_options_msg(types):
+        if dict_field:
+            for key in dict_field:
+                value = dict_field[key]
+                if value in ['', None]:
+                    raise ValueError(
+                        f"'{key}' in 'default_resolution' has no value. "
+                        f"Expected format: 'days,hours' (e.g. '{DEFAULT_TIME_RESOLUTION}')."
+                    )
```
Intended behavior

anthbel commented on grievance_social_protection/models.py:
```diff
@@ -109,6 +116,32 @@ def clean(self):
+    def filter_queryset(cls, queryset=None):
+        if queryset is None:
+            queryset = cls.objects.all()
+        queryset = queryset.filter(is_deleted=False)
```
Not relevant here.

anthbel commented:
@weilu Ready foi review.

weilu reviewed (approved):
LGTM

anthbel committed:
Fix regression
